Add support of IPv4/IPv6 DHCP.

cathay4t opened the pull request:
The sample output of `nmstatectl show` for DHCP:

    "ipv4": {
        "address": [
            {
                "ip": "192.0.2.251"
                "prefix-length": 24
            }
        ],
        "dhcp": true,
        "enabled": true
    },
    "ipv6": {
        "address": [
            {
                "ip": "fe80::1",
                "prefix-length": 64
            },
            {
                "ip": "2001:db8:1::1",
                "prefix-length": 64
            }
        ],
        "autoconf": true,
        "dhcp": true,
        "enabled": true
    },

Will raise InvalidIPv6OptionError exception if user are asking
IPv6 'dhcp: false' and 'autoconf: true'.

cathay4t commented:
Didn't have test case yet.

EdDev commented:
> Will raise InvalidIPv6OptionError exception if user are asking
> IPv6 'dhcp: false' and 'autoconf: true'.

Is it intentionally? The two are suppose to be independent.

cathay4t commented:
> > Will raise InvalidIPv6OptionError exception if user are asking
> > IPv6 'dhcp: false' and 'autoconf: true'.
> 
> Is it intentionally? The two are suppose to be independent.

It just don't supported by NetworkManager. I guess I should change it to `NoSupportOptionError`.

cathay4t commented:
Some backgrounds of NM, from `NetworkManager/libnm-util/nm-setting-ip6-config.h`:

 * 'auto' is router advertisements, DHCP, or PPP or some other device-specific manner.
 * 'dhcp' is DHCPv6 only.
 * No router advertisements mode.

EdDev commented on libnmstate/netapplier.py:
```diff
@@ -164,6 +165,20 @@ def sanitize_ethernet_state(ifaces_desired_state, ifaces_current_state):
 
+def sanitize_dhcp_state(ifaces_desired_state):
+    """
+    If DHCP enabled, if desired state does not contain any IP address,
```
What about autoconf?
Perhaps it should be something like this:
`When dynamic IP is enabled and the address missing, set an empty address list. This assures that the desired state is not complemented by the current state address values.`

EdDev commented on libnmstate/netapplier.py:
```diff
@@ -164,6 +165,20 @@ def sanitize_ethernet_state(ifaces_desired_state, ifaces_current_state):
+    for ifname, iface_state in six.viewitems(ifaces_desired_state):
+        for family in ('ipv4', 'ipv6'):
+            ip = iface_state.get(family, {})
+            if ip.get('enabled'):
```
Did you mean `dhcp` and `autoconf`?
You should not set it for non `dynamic settings.

EdDev commented on libnmstate/netapplier.py:
```diff
@@ -388,11 +407,38 @@ def _remove_iface_ipv6_link_local_addr(desired_state, current_state):
+            'dhcp': False,
+        }
+    }
+    desired_state = _dict_update(copy.deepcopy(base_conf), desired_state)
```
Why do we need to do this? For the desired state you have already done this in the `sanitize_dhcp_state` func.
Can't we do the same here for the current state?
Feels like `_remove_iface_dhcp_manual_addr` is also in this game.

Same for the IPv6 one.

EdDev commented on libnmstate/nm/ipv6.py:
```diff
@@ -57,24 +74,45 @@ def create_setting(config, base_con_profile):
-        else:
+                nmclient.NM.SETTING_IP6_CONFIG_METHOD_AUTO)
+            if config.get('address', ()):
+                logging.warning('IPv6 addresses are ignored when DHCP enabled')
```
... when dhcp/autoconf enabled

Or 'dynamic ip enabled'.

EdDev commented on libnmstate/nm/ipv6.py:
```diff
@@ -57,24 +74,45 @@ def create_setting(config, base_con_profile):
+                logging.warning('IPv6 addresses are ignored when DHCP enabled')
+        elif not is_dhcp and is_autoconf:
+            raise NoSupportIpv6DHCPOptionError(
+                'Enable autoconf only is not supported yet')
```
Perhaps: 'Autoconf without DHCP is not supported yet'

EdDev commented on libnmstate/nm/ipv6.py:
```diff
@@ -57,24 +74,45 @@ def create_setting(config, base_con_profile):
+        elif not is_dhcp and is_autoconf:
+            raise NoSupportIpv6DHCPOptionError(
+                'Enable autoconf only is not supported yet')
+        else:
```
I already lost track of all these if-else conditions. Could you please separate into functions so it will be easier to read?

cathay4t commented:
V2 patch uploaded.

Changes since V1:
 * Shoud IPv4/IPv6 as enabled when DHCP enabled even got none IP address.
 * Rework on the functions of sanitizing dhcp state when apply() as feedback.
 * Integration test attached.
 * Added a DHCP server(dnsmasq) docker instance, in order to support testing DHCP in integration test.

cathay4t commented:
V3 patch set uploaded.

Changes fromV2:
 * Use `veth` instead of new docker run for integration test.

cathay4t commented:
Rebased.

tyll commented:
ci test please

cathay4t commented on libnmstate/netapplier.py:
```diff
@@ -54,6 +55,8 @@ def _apply_ifaces_state(interfaces_desired_state, verify_change):
 
     ifaces_desired_state = sanitize_ethernet_state(ifaces_desired_state,
                                                    ifaces_current_state)
+    for iface_state in ifaces_desired_state.values():
```
python 2.7 support `dict.values()`.

cathay4t commented on automation/run-integration-tests.sh:
```diff
@@ -95,6 +95,23 @@ function open_shell {
     run_exit
 }
 
+function start_dhcpd {
```
Done.

cathay4t commented on libnmstate/netapplier.py:
```diff
@@ -150,6 +153,23 @@ def sanitize_ethernet_state(ifaces_desired_state, ifaces_current_state):
+        ip = iface_state.get(family, {})
+        if ip.get('enabled'):
+            if ip.get('dhcp') or ip.get('autoconf'):
+                if show_log and ip.get('address'):
```
It just suspend unnecessary warning message when sanitizing current state, we should only
complain with warning for desired state.

cathay4t commented:
V4 patch uploaded.

Changes since V3:
 * Change DHCP veth creation/remove function to pytest fixture.
 * Trivial fixes based on feedback.

EdDev commented on libnmstate/nm/ipv4.py:
```diff
@@ -77,3 +99,14 @@ def get_info(active_connection):
+    Get NMSettingIP4Config from NMActiveConnection.
+    For any error, return None.
+    """
+    remote_conn = active_connection.get_connection()
```
You probably should use https://lazka.github.io/pgi-docs/NM-1.0/classes/ActiveConnection.html#NM.ActiveConnection.get_dhcp4_config

We need to report the *actual* config, not the setting one (which is a desired one from NM perspective).
Could you please confirm that these are the actual and not the setting ones? I'm already confused between the active-connection and the device getters (I assumed it is the same at some point in time)

cathay4t replied:
`active_connection.get_dhcp4_config()` is getting the succeeded DHCP information.
For non-going DHCP or failed DHCP, you get `None`. We still need to check profile for that.

EdDev commented on libnmstate/nm/ipv4.py:
```diff
@@ -60,8 +67,23 @@ def get_info(active_connection):
+
     ip4config = active_connection.get_ip4_config()
     if ip4config is None:
+        # When DHCP is enable, it might be possible, the active_connection does
```
I hope that active_connection.get_dhcp4_config() should be enough to determine that IPv4 is enabled.
Then both can set the `enabled` value.

cathay4t replied:
`active_connection.get_dhcp4_config()` is getting the succeeded DHCP information.
For non-going DHCP or failed DHCP, you get `None`. We still need to check profile for that.

EdDev commented on tests/integration/dhcp_test.py:
```diff
@@ -0,0 +1,154 @@
+        except FileNotFoundError:
+            pass
+
+    request.addfinalizer(fin)
```
You should use `yield` instead of finalizer:
```
def myfixure():
    <do some setup>
    yield
    <do some teardown>
```

The teardown will occur even if during the `yield` part there is an exception.
If the exception occurs before the `yield`, you will need to wrap it with try/finally.

cathay4t replied:
The pytest suggest using the finalizer when need to handle exception during setup:

https://docs.pytest.org/en/latest/fixture.html#fixture-finalization-executing-teardown-code

And the finalizer need less line than `try+except+yield`.

EdDev replied:
I think that is an old text. You do not need try/except arround the yield.

cathay4t commented on libnmstate/nm/ipv4.py:
```diff
@@ -77,3 +99,14 @@ def get_info(active_connection):
     return info
+
+
+def _get_setting_ip_conf(active_connection):
```
I tried to align with `NMSettingIPConfig` class name.

EdDev replied:
Classes are nouns, functions are verbs.

cathay4t replied:
OK. You will get the `_get_ip_conf_setting` name.

EdDev commented on tests/integration/conftest.py:
```diff
@@ -42,6 +42,11 @@ def eth2_down():
 
 
+@pytest.fixture(scope='function')
+def eth3_down():
```
Are these `eth3` leftovers?

EdDev commented on tests/integration/dhcp_test.py:
```diff
@@ -0,0 +1,154 @@
+
+    assert libcmd.exec_cmd(['ip', 'link', 'add', DHCP_SRV_NIC, 'type', 'veth',
+                            'peer', 'name', DHCP_CLI_NIC])[0] == 0
+    fd = open(DNSMASQ_CONF_PATH, 'w')
```
Please use `with open(...) as f:`.
Same on line 77.

EdDev commented on tests/integration/dhcp_test.py:
```diff
@@ -0,0 +1,154 @@
+
+    request.addfinalizer(fin)
+
+    assert libcmd.exec_cmd(['ip', 'link', 'add', DHCP_SRV_NIC, 'type', 'veth',
```
What is the point of the assert on each execution?

cathay4t replied:
So that it could trigger clean up if any of these command fails.

EdDev commented on automation/run-integration-tests.sh:
```diff
@@ -109,4 +113,5 @@ dump_network_info
 install_nmstate
 add_extra_networks
 dump_network_info
+install_dhcpd
```
This should be part of the docker image.
But I guess we can change it later.

cathay4t replied:
Without it, the CI of this PR will fail, so we could change that later.

EdDev commented on tests/integration/dhcp_test.py:
```diff
@@ -0,0 +1,154 @@
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+#
+import os
+import pytest
```
Please separate pytest (3rd party) from the `os` std lib using a newline.

EdDev commented on tests/integration/dhcp_test.py:
```diff
@@ -0,0 +1,154 @@
+from libnmstate import nm
+
+from .testlib import assertlib
+from .testlib import statelib
```
Please sort this block.

EdDev commented on tests/integration/dhcp_test.py:
```diff
@@ -0,0 +1,154 @@
+    dhcp_cli_desired_state['ipv4']['enabled'] = True
+    dhcp_cli_desired_state['ipv4']['dhcp'] = True
+    dhcp_cli_desired_state['ipv6']['enabled'] = True
+    netapplier.apply(desired_state)
```
Please separate using a newline between the setup preparation, execution and assertion/check.

EdDev commented on tests/integration/dhcp_test.py:
```diff
@@ -0,0 +1,154 @@
+
+
+@pytest.fixture(scope='module')
+def start_dhcpd(request):
```
Please wrap the commands using functions which describe what the commands do:

- create_veth (probably setting its link up as well)
- Set addresses.
- Bind dnsmasq on a nic.
etc...

It should read nicely leaving the details deep in.

cathay4t commented:
V4 patch uploaded.

Changes since V3:
 * Decorative changes suggested in feedback.

cathay4t commented:
Added new patch to remove network requirement for integration test.
This is included in this PR in order to pass the CI test.

cathay4t commented:
@EdDev 
> Ohh, Isn't this a limitation of NM that we should resolve? I mean, it is a matter of states: One in which dhcp is down, then it is pending response, etc...
Can we open a BZ against NM on this? Perhaps you can raise it with the team first.

I don't think that's a bug, IMHO the NMDhcpConfig is intend to hold the DHCP results, it's OK to be None when pending or failed. Maybe you could ask NM dev or report bug to state your ideas there.

cathay4t commented on libnmstate/nm/ipv6.py:
```diff
@@ -15,20 +15,43 @@
+    info['dhcp'] = False
+    info['autoconf'] = False
+
+    ip_profile = nm_connection.get_ip6_profile(active_connection)
```
As we call NMRemoteConnection as a profile to indicate that's on disk configuration, I think
ip profile might better to indicate a IP config on disk.

EdDev replied:
If you love it, keep it. :)

cathay4t commented on tests/integration/dhcp_test.py:
```diff
@@ -0,0 +1,173 @@
+    assertlib.assert_state(desired_state)
+
+
+@pytest.mark.xfail(raises=nm.ipv6.NoSupportIpv6DHCPOptionError)
```
No bug reported yet and I don't intend to do so unless we have use case where 'autoconf' only
is needed.

EdDev commented on libnmstate/netapplier.py:
```diff
@@ -150,6 +152,21 @@ def sanitize_ethernet_state(ifaces_desired_state, ifaces_current_state):
+    list. This assures that the desired state is not complemented by the
+    current state address values.
+    """
+    print(ifaces_state)
```
Left over

EdDev commented on libnmstate/netapplier.py:
```diff
@@ -329,6 +346,8 @@ def assert_ifaces_state(ifaces_desired_state, ifaces_current_state):
         iface_dstate, iface_cstate = _remove_iface_ipv6_link_local_addr(
             iface_dstate, iface_cstate)
+        iface_cstate = sanitize_dhcp_state(
+            {iface_cstate['name']: iface_cstate})[iface_cstate['name']]
```
Can we use `ifname` instead of `iface_cstate['name']`? Not a must, it is just less noisy.

EdDev commented on libnmstate/nm/connection.py:
```diff
@@ -164,3 +164,25 @@ def _logging_connection_info(con_setting, source):
     )
+
+
+def get_ip4_profile(active_connection):
```
I think that routes can depend on the ipv4 module.
I think it will be cleaner if the connection module will not know about the setting details, passing that knowledge to the other modules. Otherwise you will get into this module all setting getters from a lot of sources.

EdDev commented:
ci test please
